tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/AUC.java

```diff
@@ -0,0 +1,1021 @@
+   */
+  public AUC(
+      Ops tf,
+      String name,
+      Integer numThresholds,
+      AUCCurve curve,
+      AUCSummationMethod summationMethod,
+      float[] thresholds,
+      boolean multiLabel,
+      Operand<T> labelWeights,
+      long seed,
+      Class<T> type) {
+    super(tf, name == null ? DEFAULT_NAME : name, seed);
+    this.truePositivesName = this.getVariableName(TRUE_POSITIVES);
+    this.falsePositivesName = this.getVariableName(FALSE_POSITIVES);
+    this.trueNegativesName = this.getVariableName(TRUE_NEGATIVES);
+    this.falseNegativesName = this.getVariableName(FALSE_NEGATIVES);
+    this.curve = curve;
+    this.summationMethod = summationMethod;
+    this.type = type;
+
+    this.multiLabel = multiLabel;
+
+    if (thresholds != null) { // ignore numThresholds
+      for (float t : thresholds)
+        if (t < 0.0f || t > 1.0f)
+          throw new IllegalArgumentException(
+              String.format(
+                  "Threshold values must be in [0, 1]. Invalid values: %s",
+                  Arrays.toString(thresholds)));
+      this.numThresholds = thresholds.length + 2;
+      Arrays.sort(thresholds);
+    } else {
+      if (numThresholds <= 1) throw new IllegalArgumentException("numThresholds must be > 1.");
+      this.numThresholds = numThresholds;
+      thresholds = new float[numThresholds - 2];
+      // linearly interpolate (numThresholds - 2) thresholds between endpoints
+      for (int i = 0; i < thresholds.length; i++) {
+        thresholds[i] = (i + 1) * 1.0f / (this.numThresholds - 1);
+      }
+    }
+    // Add an endpoint "threshold" below zero and above one for either
+    // threshold method to account for floating point imprecision.
+    if (thresholds.length != this.numThresholds - 2)
+      throw new IllegalArgumentException(
+          "Thresholds length must contain numThresholds - 2 entries");
+    this.thresholds = new float[this.numThresholds];
+    this.thresholds[0] = -EPSILON;
+    System.arraycopy(thresholds, 0, this.thresholds, 1, thresholds.length);
+    this.thresholds[this.numThresholds - 1] = 1 + EPSILON;
+
+    if (labelWeights != null) {
+      // assert that labelWeights are non-negative.
+
+      this.labelWeights = labelWeights;
+      Op checks =
+          getTF()
+              .withSubScope("AUC")
+              .assertThat(
+                  getTF()
+                      .math
+                      .greaterEqual(
+                          labelWeights, cast(getTF(), getTF().constant(0), labelWeights.type())),
+                  Collections.singletonList(
+                      getTF().constant("All values of labelWeights must be non-negative.")));
+
+      Ops ltf =
+          getTF()
+              .withSubScope("updateState")
+              .withControlDependencies(Collections.singletonList(checks));
+
+      this.labelWeights = ltf.identity(this.labelWeights);
+    }
+
+    if (this.multiLabel) {
+      this.numLabels = null;
```

Comment:
What's the purpose of setting `numLabels` to `null` here? I don't see the Python code doing that. It doesn't matter in our code -- we won't access `numLabels` again. But it may trip us up later, such as if we implement a counterpart to Python's `reset_states`.

Reply:
The condition of multilLabel + labelWeights != null is handled in 
`AUC.result()`, not in `MetricsHelper.updateConfusionMatrixVariables()`.

However, you are right that `this.labelWeights` should not be changed as it is used later. I will fix the code to pass null to `updateConfusionMatrixVariables` if `multiLabel` is true.

```
MetricsHelper.updateConfusionMatrixVariables(
            tf,
            confusionMatrix,
            varInitializers,
            tLabels,
            tPredictions,
            tf.constant(thresholds),
            null,
            null,
            tSampleWeights,
            isMultiLabel(),
            isMultiLabel() ? null : labelWeights));
```

Reply:
👍 done